File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.cc
#include <iostream>
#include <queue>
#include <tuple>

using namespace std;

int main(void)
{
    int N, M;
    cin >> N >> M;

    int x, y;
    char MAP[50][50] = {0};

    for (int i = 0; i < N; i++)
    {
        cin >> MAP[i];
        for (int j = 0; j < M; j++)
        {
            if (MAP[i][j] == '0')
            {
                x = i;
                y = j;
                MAP[i][j] = '.';
            }
        }
    }

    const int DXs[4] = {-1, 1, 0, 0};
    const int DYs[4] = {0, 0, -1, 1};

    bool visited[50][50][0b1000000] = {0};
    visited[x][y][0] = true;

    typedef tuple<int, int, int, int> qelem;
    queue<qelem> q;
    q.push(make_tuple(x, y, 0, 0));

    while (!q.empty())
    {
        qelem now = q.front();

        int x = get<0>(now);
        int y = get<1>(now);
        int keys = get<2>(now);
        int level = get<3>(now);

        q.pop();

        for (int i = 0; i < 4; i++)
        {
            int nx = x + DXs[i];
            int ny = y + DYs[i];

            if (!(0 <= nx && nx < N && 0 <= ny && ny < M))
                continue;

            if (visited[nx][ny][keys] or MAP[nx][ny] == '#')
                continue;

            else if (MAP[nx][ny] == '1')
            {
                x = nx;
                y = ny;
                cout << level + 1 << endl;
                return 0;
            }
            if (MAP[nx][ny] == '.' && !visited[nx][ny][keys])
            {
                visited[nx][ny][keys] = true;
                q.push(make_tuple(nx, ny, keys, level + 1));
            }
            else if ('A' <= MAP[nx][ny] && MAP[nx][ny] <= 'Z')
            {
                if ((keys & (1 << (MAP[nx][ny] - 'A'))) != 0 && !visited[nx][ny][keys])
                {
                    visited[nx][ny][keys] = true;
                    q.push(make_tuple(nx, ny, keys, level + 1));
                }
            }
            else if ('a' <= MAP[nx][ny] && MAP[nx][ny] <= 'z')
            {
                int tmp_keys = keys | (1 << (MAP[nx][ny] - 'a'));
                if (!visited[nx][ny][keys])
                {
                    visited[nx][ny][tmp_keys] = true;
                    q.push(make_tuple(nx, ny, tmp_keys, level + 1));
                }
            }
        }
    }

    cout << -1 << endl;

    return 0;
}

File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.py
from collections import deque

N, M = map(int, input().split())
MAP = [input() for _ in range(N)]

START = (0, 0)
for i in range(N):
    for j in range(M):
        if MAP[i][j] == '0':
            START = (i, j)


x = START[0]
y = START[1]


def bfs():
    global x, y

    VISITED = [[[False for _ in range(0b1000000)]
                for _ in range(M)] for _ in range(N)]

    VISITED[x][y][0] = True

    dxs = [-1, 1, 0, 0]
    dys = [0, 0, -1, 1]

    queue = deque([(x, y, 0, 0)])

    while queue:
        x, y, keys, level = queue.popleft()

        for dx, dy in zip(dxs, dys):
            nx = x + dx
            ny = y + dy
            if not (0 <= nx < N and 0 <= ny < M):
                continue
            if VISITED[nx][ny][keys] or MAP[nx][ny] == '#':
                continue
            elif MAP[nx][ny] == '1':
                x, y = nx, ny
                return level + 1
            elif MAP[nx][ny] in ['.', '0']:
                if not VISITED[nx][ny][keys]:
                    VISITED[nx][ny][keys] = True
                    queue.append((nx, ny, keys, level + 1))
            elif MAP[nx][ny].isupper():
                if keys & (1 << (ord(MAP[nx][ny].lower()) - ord('a'))) != 0:
                    if not VISITED[nx][ny][keys]:
                        VISITED[nx][ny][keys] = True
                        queue.append((nx, ny, keys, level + 1))
            elif MAP[nx][ny].islower():
                tmp_keys = keys | (1 << (ord(MAP[nx][ny]) - ord('a')))
                if not VISITED[nx][ny][tmp_keys]:
                    VISITED[nx][ny][tmp_keys] = True
                    queue.append((nx, ny, tmp_keys, level + 1))

    return -1


answer = bfs()
print(answer)
